Testing with various versions of pip

It seems that right now we have only one version of pip in our testing dependencies and moreover this version is not locked. If I am not wrong, the version of pip we are using in tests is the one which virtualenv bundles and it's therefore available in pytest-venv fixture where our test run. Right?

IMHO, this has to be extended if we want to keep backward compatibility and test micropipenv with the latest pip before it's bundled in virtualenv.

What do you think about it? I can try to prepare something. A solution also has to be optional so the tests won't require internet.

> I was wondering about creating tests that would create matrix `Python version` x `pip version` so we test against these two (not limiting to s2i case). Having just virtualenv might be good for s2i case as implemented now. Do you plan to use Python's venv module in future s2i container images?

I don't understand here. Virtualenv is used heavily in s2i-python-container. Newer versions with Python >= 3.6 use venv from Python stdlib and older versions use python-virtualenv RPM package.

> BTW to have proper tests, we would also include toml as a dependency but I think we could ignore it (at least for now). toml might be soon incorporated into Python's standard library, but we'll need to be backwards compatible.

Different versions of toml might be included in the same way as pip, see below.

I was thinking about code like this:

```
import os

def test_install(venv):
    pip_version = os.getenv("PIP_VERSION", None)
    if pip_version is not None:
        venv.install(f"pip{pip_version}")
    print(venv.get_version("pip"))
    assert False
```
The first three lines would be implemented somehow in each test which uses venv fixture. 

So, the virtualenv we are testing micropipenv in will get the default pip (bundled in virtualenv or from distro in case of Fedora) as it does now without any need for internet access. 

```
python -m pytest test_example.py
==================================== test session starts =====================================
platform linux -- Python 3.9.0b1, pytest-5.4.2, py-1.8.1, pluggy-0.13.1
rootdir: /home/lbalhar/.virtualenvs/tmp-8c0a1522bd3ca2f
plugins: venv-0.2
collected 1 item                                                                             

test_example.py F                                                                      [100%]

========================================== FAILURES ==========================================
________________________________________ test_install ________________________________________

venv = <pytest_venv.VirtualEnvironment object at 0x7fbaba7a7b80>

    def test_install(venv):
        pip_version = os.getenv("PIP_VERSION", None)
        if pip_version is not None:
            venv.install(f"pip=={pip_version}")
        print(venv.get_version("pip"))
>       assert False
E       assert False

test_example.py:8: AssertionError
----------------------------------- Captured stdout setup ------------------------------------
created virtual environment CPython3.9.0.beta.1-64 in 122ms
  creator CPython3Posix(dest=/tmp/pytest-of-lbalhar/pytest-2/test_install0, clear=False, global=False)
  seeder FromAppData(download=False, pip=latest, setuptools=latest, wheel=latest, via=copy, app_data_dir=/home/lbalhar/.local/share/virtualenv/seed-app-data/v1.0.1)
  activators BashActivator,CShellActivator,FishActivator,PowerShellActivator,PythonActivator,XonshActivator
------------------------------------ Captured stdout call ------------------------------------
20.1.1
================================== short test summary info ===================================
FAILED test_example.py::test_install - assert False
===================================== 1 failed in 0.38s ======================================
```

But if you specify `PIP_VERSION` environment variable, the specific version will be installed and micropipenv will be tested with it.

```
PIP_VERSION="==19.3.1" python -m pytest test_example.py
==================================== test session starts =====================================
platform linux -- Python 3.9.0b1, pytest-5.4.2, py-1.8.1, pluggy-0.13.1
rootdir: /home/lbalhar/.virtualenvs/tmp-8c0a1522bd3ca2f
plugins: venv-0.2
collected 1 item                                                                             

test_example.py F                                                                      [100%]

========================================== FAILURES ==========================================
________________________________________ test_install ________________________________________

venv = <pytest_venv.VirtualEnvironment object at 0x7f49bc1d00d0>

    def test_install(venv):
        pip_version = os.getenv("PIP_VERSION", None)
        if pip_version is not None:
            venv.install(f"pip=={pip_version}")
        print(venv.get_version("pip"))
>       assert False
E       assert False

test_example.py:8: AssertionError
----------------------------------- Captured stdout setup ------------------------------------
created virtual environment CPython3.9.0.beta.1-64 in 105ms
  creator CPython3Posix(dest=/tmp/pytest-of-lbalhar/pytest-3/test_install0, clear=False, global=False)
  seeder FromAppData(download=False, pip=latest, setuptools=latest, wheel=latest, via=copy, app_data_dir=/home/lbalhar/.local/share/virtualenv/seed-app-data/v1.0.1)
  activators BashActivator,CShellActivator,FishActivator,PowerShellActivator,PythonActivator,XonshActivator
------------------------------------ Captured stdout call ------------------------------------
Collecting pip==19.3.1
  Using cached pip-19.3.1-py2.py3-none-any.whl (1.4 MB)
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 20.1.1
    Uninstalling pip-20.1.1:
      Successfully uninstalled pip-20.1.1
Successfully installed pip-19.3.1
19.3.1
================================== short test summary info ===================================
FAILED test_example.py::test_install - assert False
===================================== 1 failed in 2.36s ======================================
```

And then, we can use the capabilities of tox to create the matrix you mentioned combining Python versions and pip versions together:

```
[tox]
envlist = py{36,37,38,39}-pip{190,200,201}
skipsdist = True

[testenv]
commands = pytest -s -vv test_example.py
setenv =
    pip190: PIP_VERSION = >=19.0,<20.0
    pip200: PIP_VERSION = >=20.0,<20.1
    pip201: PIP_VERSION = >=20.1,<20.2
deps =
    pytest
    pytest-venv
```

Which will lead to twelve testing environments:
```
py36-pip190
py36-pip200
py36-pip201
py37-pip190
py37-pip200
py37-pip201
py38-pip190
py38-pip200
py38-pip201
py39-pip190
py39-pip200
py39-pip201
```

What do you think about it? If it sounds good to you, I can try to implement it into micropipenv.

--- a/tests/test_micropipenv.py
+++ b/tests/test_micropipenv.py
@@ -33,6 +33,22 @@ import micropipenv
 
 
 _DATA_DIR = os.path.abspath(os.path.join(os.path.dirname(os.path.relpath(__file__)), "data"))
+# Version of pip to test micropipenv with
+# the default is the pip wheel bundled in virtualenv package
+PIP_VERSION = os.getenv("PIP_VERSION")
+
+
+@pytest.fixture(name="venv")
+def venv_with_pip(venv):
+    """Fixture for virtual environment with specific version of pip.
+
+    Fixture uses the original one from pytest_venv,
+    installs pip if PIP_VERSION is given
+    and overwrites the original fixture name
+    """
+    if PIP_VERSION is not None:
+        venv.install(f"pip{PIP_VERSION}")
+    yield venv
 
 
 @contextmanager

--- a/tox.ini
+++ b/tox.ini
@@ -1,5 +1,5 @@
 [tox]
-envlist = py36,py37,py38,py39
+envlist = py{36,37,38,39}-pip{190,200,201}
 skipsdist = True
 
 [testenv]
@@ -12,3 +12,7 @@ deps =
     pytest-timeout
     pytest-venv
     toml
+setenv =
+    pip190: PIP_VERSION = >=19.0,<20.0
+    pip200: PIP_VERSION = >=20.0,<20.1
+    pip201: PIP_VERSION = >=20.1,<20.2

